model/policy.go

```diff
@@ -122,59 +138,69 @@ func (model Model) HasPolicies(sec string, ptype string, rules [][]string) bool
 }
 
 // AddPolicy adds a policy rule to the model.
-func (model Model) AddPolicy(sec string, ptype string, rule []string) {
-	model[sec][ptype].Policy = append(model[sec][ptype].Policy, rule)
-	model[sec][ptype].PolicyMap[strings.Join(rule, DefaultSep)] = len(model[sec][ptype].Policy) - 1
+func (model *Model) AddPolicy(sec string, ptype string, rule []string) {
+	model.Mutex.Lock()
+	defer model.Mutex.Unlock()
+	model.Data[sec][ptype].Policy = append(model.Data[sec][ptype].Policy, rule)
+	model.Data[sec][ptype].PolicyMap[strings.Join(rule, DefaultSep)] = len(model.Data[sec][ptype].Policy) - 1
 }
 
 // AddPolicies adds policy rules to the model.
-func (model Model) AddPolicies(sec string, ptype string, rules [][]string) {
+func (model *Model) AddPolicies(sec string, ptype string, rules [][]string) {
+	model.Mutex.Lock()
+	defer model.Mutex.Unlock()
 	for _, rule := range rules {
 		hashKey := strings.Join(rule, DefaultSep)
-		_, ok := model[sec][ptype].PolicyMap[hashKey]
+		_, ok := model.Data[sec][ptype].PolicyMap[hashKey]
 		if ok {
 			continue
 		}
-		model[sec][ptype].Policy = append(model[sec][ptype].Policy, rule)
-		model[sec][ptype].PolicyMap[hashKey] = len(model[sec][ptype].Policy) - 1
+		model.Data[sec][ptype].Policy = append(model.Data[sec][ptype].Policy, rule)
+		model.Data[sec][ptype].PolicyMap[hashKey] = len(model.Data[sec][ptype].Policy) - 1
 	}
 }
 
 // RemovePolicy removes a policy rule from the model.
-func (model Model) RemovePolicy(sec string, ptype string, rule []string) bool {
-	index, ok := model[sec][ptype].PolicyMap[strings.Join(rule, DefaultSep)]
+func (model *Model) RemovePolicy(sec string, ptype string, rule []string) bool {
+	model.Mutex.Lock()
+	defer model.Mutex.Unlock()
+	index, ok := model.Data[sec][ptype].PolicyMap[strings.Join(rule, DefaultSep)]
 	if !ok {
 		return false
 	}
 
-	model[sec][ptype].Policy = append(model[sec][ptype].Policy[:index], model[sec][ptype].Policy[index+1:]...)
-	delete(model[sec][ptype].PolicyMap, strings.Join(rule, DefaultSep))
-	for i := index; i < len(model[sec][ptype].Policy); i++ {
-		model[sec][ptype].PolicyMap[strings.Join(model[sec][ptype].Policy[i], DefaultSep)] = i
+	model.Data[sec][ptype].Policy = append(model.Data[sec][ptype].Policy[:index], model.Data[sec][ptype].Policy[index+1:]...)
+	delete(model.Data[sec][ptype].PolicyMap, strings.Join(rule, DefaultSep))
+	for i := index; i < len(model.Data[sec][ptype].Policy); i++ {
+		model.Data[sec][ptype].PolicyMap[strings.Join(model.Data[sec][ptype].Policy[i], DefaultSep)] = i
```

Comment:
It looks like an expensive operation, I'm wondering why dont we implement a class hashset using map.

Or use something like swap remove,  swap the policy with the last one and then remove only the last one. this way we avoid touching the others

Reply:
The problem is that the result of casbin enforce is affected by the order of the policy, so we must store the order of the policy through slices, and this map is to speed up querying the policy position. If we want to change to the classic hashset, we need to give up this feature.

And this operation is to update the data in the map after removing the policy, so we must modify the position of each policy in the map one by one

Reply:
We can use linked hash map to keep order and implement a hashset with it.  I believe it solves this problem. 

Also even we insist on this solution, we need to use at least a hash function instead if Join directly the policy which can be heavy to store.  It doubles the used memory

Reply:
BTW, 

> The problem is that the result of casbin enforce is affected by the order of the policy, so we must store the order of the policy through slices, and this map is to speed up querying the policy position. If we want to change to the classic hashset, the result of the enforcer will be uncertain.


when we are enforcing, there isn't any change on policy and policy order (thanks lock).  I still think we can do swap remove instead of simple remove. This drops down from O(n) to O(1)

Reply:
For swap remove，suppose we have such a policy
```
g, alice, data1_deny_group

p, bob, data2, read, deny
p, alice, data1, read, allow
p, data1_deny_group, data1, write, deny
p, data1_deny_group, data1, read, deny
// enforce(alice, data1, read) =>  true
```

If we call `remove(bob, data2, deny)`, then it should become

```
p, data1_deny_group, data1, read, deny
p, alice, data1, read, allow
p, data1_deny_group, data1, write, deny
// enforce(alice, data1, read) =>  false
```
I think it’s not good to remove one policy but affect the other

Reply:
I remember that we once mentioned the linked hashmap when doing this https://github.com/casbin/casbin/pull/500, we should listen to the opinions of other people @hsluoyz @nodece  @divy9881 

I totally agree with storing the hash value, I will add it in another pr

Reply:
It is very important to explore a new structure to store policy, we can implement the feature in next PR.

Reply:
then we need a real hash set which keeps order. Go should have one.